### PR TITLE
MVC

### PR DESCRIPTION
**Pull request by IgorPrihodko**

### CODE REVIEW

File: src/main/java/com/prykhodko/shop/controller/ProductController.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        this.productService = productService;
+    }
+
+    @GetMapping("/products")
+    public String allProducts(ModelMap modelMap) {
+        modelMap.addAttribute("products", productService.getAll());
+        return "products";
+    }
+
+    @GetMapping("/addProduct")
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/controller/ProductController.java]: better: "/product/add"

File: src/main/java/com/prykhodko/shop/controller/ProductController.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                          @RequestParam(value = "price", required = false) BigDecimal price,
+                          ModelMap modelMap) {
+        if (title.isEmpty() || description.isEmpty() || price.equals(BigDecimal.ZERO)) {
+            modelMap.addAttribute("wrong", "Wrong data entered");
+            return "redirect:addProduct";
+        }
+
+        Product product = new Product(title, description, price);
+        productService.addProduct(product);
+        logger.warn("Add new product " + product + " to db");
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/controller/ProductController.java]: maybe better info?

File: src/main/java/com/prykhodko/shop/controller/StartPageController.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    private final UserService userService;
+    private final ProductService productService;
+
+    @Autowired
+    public StartPageController(UserService userService, ProductService productService) {
+        this.userService = userService;
+        this.productService = productService;
+    }
+
+    @GetMapping("/")
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/controller/StartPageController.java]: try to use @postConstract

File: src/main/java/com/prykhodko/shop/controller/UserController.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.SessionAttributes;
+
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Optional;
+
+@Controller
+@RequestMapping("/admin")
+@SessionAttributes({"user", "basket", "wrong"})
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/controller/UserController.java]: the same

File: src/main/java/com/prykhodko/shop/dao/impl/BasketDaoHibernate.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    @Autowired
+    public BasketDaoHibernate(SessionFactory sessionFactory) {
+        this.sessionFactory = sessionFactory;
+    }
+
+    @Override
+    public void addBasket(Basket basket) {
+        Session session = sessionFactory.getCurrentSession();
+        session.save(basket);
+        logger.warn("added new basket " + basket);
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/dao/impl/BasketDaoHibernate.java]: info vs warn???

File: src/main/java/com/prykhodko/shop/service/impl/ConfirmationCodeServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    private final ConfirmationCodeDao confirmationCodeDao;
+
+    @Autowired
+    public ConfirmationCodeServiceImpl(ConfirmationCodeDao confirmationCodeDao) {
+        this.confirmationCodeDao = confirmationCodeDao;
+    }
+
+    @Override
+    @Transactional
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/service/impl/ConfirmationCodeServiceImpl.java]: we use Transactional over the class, so we don't need it here

File: src/main/java/com/prykhodko/shop/service/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    private final UserDao userDao;
+
+    @Autowired
+    public UserServiceImpl(UserDao userDao) {
+        this.userDao = userDao;
+    }
+
+    @Override
+    @Transactional
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/service/impl/UserServiceImpl.java]: the same

File: src/main/java/com/prykhodko/shop/service/impl/ConfirmationCodeServiceImpl.java (hunk 6 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/service/impl/ConfirmationCodeServiceImpl.java]: fixed

File: src/main/java/com/prykhodko/shop/dao/impl/BasketDaoHibernate.java (hunk 5 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/dao/impl/BasketDaoHibernate.java]: fixed

File: src/main/java/com/prykhodko/shop/controller/ProductController.java (hunk 2 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/controller/ProductController.java]: fixed

File: spring4-hibernate5-example.iml [hunk 8]
```diff
@@ -4,11 +4,19 @@
     <facet type="Spring" name="Spring">
```
[the remark this answers is not in the document]
Comment (IgorPrihodko) [reply spring4-hibernate5-example.iml]: fixed

File: src/main/java/com/prykhodko/shop/service/impl/ProductServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+    private final ProductDao productDao;
+
+    @Autowired
+    public ProductServiceImpl(ProductDao productDao) {
+        this.productDao = productDao;
+    }
+
+    @Override
+    @Transactional
```
[the remark this answers is not in the document]
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/service/impl/ProductServiceImpl.java]: fixed

File: src/main/java/com/prykhodko/shop/service/impl/UserServiceImpl.java (hunk 7 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/service/impl/UserServiceImpl.java]: fixed

File: src/main/java/com/prykhodko/shop/utils/TotalPriceCounter.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
+package com.prykhodko.shop.utils;
+
+import com.prykhodko.shop.model.Basket;
+import com.prykhodko.shop.model.Product;
+
+import java.math.BigDecimal;
+
+public class TotalPriceCounter {
+
+    public static BigDecimal count(Basket basket) {
```
[the remark this answers is not in the document]
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/utils/TotalPriceCounter.java]: fixed

File: src/main/java/com/prykhodko/shop/controller/UserController.java (hunk 4 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/controller/UserController.java]: fixed

File: src/main/java/com/prykhodko/shop/controller/StartPageController.java (hunk 3 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/controller/StartPageController.java]: done

File: src/main/java/com/prykhodko/shop/controller/OrderController.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.SessionAttributes;
+
+import java.math.BigDecimal;
+
+@Controller
+@RequestMapping("/user")
+@SessionAttributes({"user", "basket", "totalPrice", "wrong"})
```
[the remark this answers is not in the document]
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/controller/OrderController.java]: fixed

File: src/main/java/com/prykhodko/shop/controller/ProductController.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/controller/ProductController.java]: info)

File: src/main/java/com/prykhodko/shop/controller/ProductController.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+        Product product = productService.getById(id).get();
+        if (title.isEmpty() || description.isEmpty() || price.equals(BigDecimal.ZERO)) {
+            modelMap.addAttribute("wrong", "Wrong data entered");
+            return "redirect:editProduct";
+        }
+        product.setTitle(title);
+        product.setDescription(description);
+        product.setPrice(price);
+        productService.updateProduct(id, product);
+        logger.warn("Edit product " + product + " in db");
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/controller/ProductController.java]: :)

File: src/main/java/com/prykhodko/shop/controller/ProductController.java [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        modelMap.addAttribute("title", deletedProduct.getTitle());
+        modelMap.addAttribute("description", deletedProduct.getDescription());
+        modelMap.addAttribute("price", deletedProduct.getPrice());
+        return "deleteProduct";
+    }
+
+    @PostMapping("/deleteProduct")
+    public String deleteProduct(@ModelAttribute(value = "id") Long id) {
+        productService.removeProduct(id);
+        logger.warn("Remove product with id " + id);
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/controller/ProductController.java]: :)

File: src/main/java/com/prykhodko/shop/dao/impl/OrderDaoHibernate.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    private final SessionFactory sessionFactory;
+
+    @Autowired
+    public OrderDaoHibernate(SessionFactory sessionFactory) {
+        this.sessionFactory = sessionFactory;
+    }
+
+    @Override
+    public void addOrder(StockOnOrder stockOnOrder) {
+        Session session = sessionFactory.getCurrentSession();
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/dao/impl/OrderDaoHibernate.java]: we can simplify:
sessionFactory.getCurrentSession().save(stockOnOrder);

File: src/main/java/com/prykhodko/shop/service/impl/MailServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+import javax.mail.PasswordAuthentication;
+import javax.mail.Session;
+import javax.mail.Transport;
+import javax.mail.internet.InternetAddress;
+import javax.mail.internet.MimeMessage;
+import java.util.Properties;
+
+public class MailServiceImpl implements MailService {
+
+    Logger logger = Logger.getLogger(MailServiceImpl.class);
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/service/impl/MailServiceImpl.java]: add access modifier

File: src/main/java/com/prykhodko/shop/utils/TotalPriceCounter.java [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+public class TotalPriceCounter {
+
+    private final ProductService productService;
+
+    @Autowired
+    public TotalPriceCounter(ProductService productService) {
+        this.productService = productService;
+    }
+
+    public BigDecimal count(Basket basket) {
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/utils/TotalPriceCounter.java]: better move this method in basketDaoImpl class

File: src/main/java/com/prykhodko/shop/utils/TotalPriceCounter.java [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+import com.prykhodko.shop.service.ProductService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.math.BigDecimal;
+
+@Service
+public class TotalPriceCounter {
+
+    private final ProductService productService;
```
Comment (Sofasmile) [inline src/main/java/com/prykhodko/shop/utils/TotalPriceCounter.java]: we shouldn't use service here

File: src/main/java/com/prykhodko/shop/controller/ProductController.java (hunk 12 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/controller/ProductController.java]: fixed

File: src/main/java/com/prykhodko/shop/controller/ProductController.java (hunk 13 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/controller/ProductController.java]: fixed

File: src/main/java/com/prykhodko/shop/controller/ProductController.java (hunk 14 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/controller/ProductController.java]: fixed

File: src/main/java/com/prykhodko/shop/dao/impl/OrderDaoHibernate.java (hunk 15 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/dao/impl/OrderDaoHibernate.java]: done

File: src/main/java/com/prykhodko/shop/service/impl/MailServiceImpl.java (hunk 16 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/service/impl/MailServiceImpl.java]: done

File: src/main/java/com/prykhodko/shop/utils/TotalPriceCounter.java (hunk 17 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/utils/TotalPriceCounter.java]: fixed

File: src/main/java/com/prykhodko/shop/utils/TotalPriceCounter.java (hunk 18 above)
Comment (IgorPrihodko) [reply src/main/java/com/prykhodko/shop/utils/TotalPriceCounter.java]: done

**Pull request merged**